pvlib/pvsystem.py

```diff
@@ -2537,6 +2553,73 @@ def pvwatts_losses(soiling=2, shading=3, snow=0, mismatch=2, wiring=2,
     return losses
 
 
+def irradiance_loss_pvsyst(effective_irradiance, irradiance_shading_loss,
+                           irradiance_snow_loss, irradiance_soiling_loss):
+    r"""
+    Calculates plane of array irradiance losses using a model based on the
+    PVSyst methodology [1]_ of linearly reducing the effective irradiance on a
+    module.  This doesn't account for the electrical losses due to irregular
+    irradiance on different solar cell strings but rather reduces the
+    irradiance input to the single diode equation.
+
+    The losses are compounded using the following equation:
+
+    .. math::
+
+        L_{total}(\%) = 100 [ 1 - \Pi_i ( 1 - \frac{L_i}{100} ) ]
+
+    Note the parameters must each be a series with a DatetimeIndex.  The
+    index of the returned series will be the same as the effective_irradiance
+    index.  All other parameters will be resampled to match this index with
+    a "fill forward" method for filling holes.
+
+    Parameters
+    ----------
+    effective_irradiance : Series
+            The irradiance (W/m2) that is to be reduced.  Plane-of-array
+            irradiance that already accounts for Angle of Incidence and
+            Spectral losses.
```

Comment:
Question for my benefit: PVsyst applies the snow, shade and soiling losses after reflection losses?

Reply:
The PVSyst near shading loss model is broken out into beam, diffuse, and albedo components.  (It's far shading loss is only a beam component.)  It totals these components and applies them as a global shading loss.  Then it applies an Incident Angle Modifier to the beam component.  Then the soiling losses (which have snow losses lumped into them) are applied as a constant haircut to effective irradiance for the entire month.  So, PVSyst is: Shading -> Incident Angle Modifier -> Soiling.

An irradiance_shading_loss_model_pvsyst will need to calculate its global shading loss by accessing the total_irrad['poa_direct'] and total_irrad['poa_diffuse'] parameters.  An alternative implementation to this PR would be to include the shading loss factors when calculating the effective_irradiance_model.

As long as the shading_loss_model properly accounts for each component of the global shading loss and applies this loss as a percentage of total irradiance (as PVSyst does), I don't believe the order of the losses affects the final result.  With this in mind, my objective was to make this PR as small as possible and apply the the shading/soiling/snow losses to the effective irradiance without changing the effective_irradiance_model in PVLib.  Though I'm certainly open to different implementation strategies.

Reply:
No objections to the approach from me, only encouragement. We can plan to add other small pieces (e.g. far shading) for steps in the PVsyst loss chain, and eventually a convenience method that strings the pieces together, to go from beam and diffuse to effective irradiance with losses.